### ISSUE TITLE
No longer works in Home Assistant 2024.12 container

### ISSUE THREAD

**Issue by klein1jo** [widewing/ha-toyota-na#101]
Running the official Home Assistant container. Version 2024.12 was recently released which requires Python 3.13. Integration on longer works.

Setup failed for custom integration 'toyota_na': Unable to import component: No module named 'toyota_na'
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/setup.py", line 334, in _async_setup_component
    component = await integration.async_get_component()
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/loader.py", line 1034, in async_get_component
    self._component_future.result()
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^
  File "/usr/src/homeassistant/homeassistant/loader.py", line 1014, in async_get_component
    comp = await self.hass.async_add_import_executor_job(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        self._get_component, True
        ^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/usr/local/lib/python3.13/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/src/homeassistant/homeassistant/loader.py", line 1074, in _get_component
    ComponentProtocol, importlib.import_module(self.pkg_path)
                       ~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/util/loop.py", line 200, in protected_loop_func
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.13/importlib/__init__.py", line 88, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1331, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 935, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 1022, in exec_module
  File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
  File "/config/custom_components/toyota_na/__init__.py", line 6, in <module>
    from toyota_na.auth import ToyotaOneAuth
ModuleNotFoundError: No module named 'toyota_na'

**Comment by va2na** [comment]
Same here

**Comment by seang96** [comment]
Can you try running startup command as below? 

`/bin/bash -c PYTHONPATH="$PYTHONPATH:/config/deps" /init`

I noticed that. HASS downloads integrations libraries to this deps folder but it is not referenced in their Environment variables, so it doesn't get used with containers. I assume this is them treating the container environments as a second class citizen again.

**Comment by klein1jo** [comment]
No success.  Ran:

`podman exec -it home-assistant /bin/bash -c PYTHONPATH="$PYTHONPATH:/config/deps" /init`

Restarted just the hass app from "Developer Tools".  Same error persists.

**Comment by seang96** [comment]
Just to make sure, /config is where your persistent data is and can you validate there is a deps folder?

Also another test you can try

`pip install toyota-na` in the container and restart home assistant from the settings in home assistant, not the container.

**Comment by klein1jo** [comment]
Yes /config/deps exists.  Here's some commands for you from in the container:

```
c92bd0e92f64:/config/deps# ls -d to*
toyota_na                  toyota_na-2.1.1.dist-info
c92bd0e92f64:/config/deps#
c92bd0e92f64:/config/deps#
c92bd0e92f64:/config/deps#
c92bd0e92f64:/config/deps# /bin/bash -c PYTHONPATH="$PYTHONPATH:/config/deps" /init
c92bd0e92f64:/config/deps#
c92bd0e92f64:/config/deps# echo $PYTHONPATH

c92bd0e92f64:/config/deps#
```

As you can see toyota-na already exists at least in my persistent data dir.  I would have expected my echo there to show a path after running the command you suggested.

I ran `pip install toyota-na`.  That worked.  I now have toyota-na in my container's `pip list` command.  So they're not including /config/deps like you suspected and that setting of PYTHONPATH didn't work as expected?

I'm assuming if they're not including /config/deps my next step is to open an issue on the hass docker project?

**Comment by finity69x2** [comment]
When I run the recommended command above ( pip install toyota-na ) I get the following output in the container console of portainer:

```
Looking in indexes: https://pypi.org/simple, https://wheels.home-assistant.io/musllinux-index/
Requirement already satisfied: toyota-na in /usr/local/lib/python3.12/site-packages (2.1.1)
Requirement already satisfied: PyJWT>=2.1.0 in /usr/local/lib/python3.12/site-packages (from toyota-na) (2.9.0)
Requirement already satisfied: aiohttp>=3.8.1 in /usr/local/lib/python3.12/site-packages (from toyota-na) (3.10.11)
Requirement already satisfied: cryptography>=35.0.0 in /usr/local/lib/python3.12/site-packages (from toyota-na) (43.0.1)
Requirement already satisfied: pytz>=2022.1 in /usr/local/lib/python3.12/site-packages (from toyota-na) (2024.2)
Requirement already satisfied: aiohappyeyeballs>=2.3.0 in /usr/local/lib/python3.12/site-packages (from aiohttp>=3.8.1->toyota-na) (2.4.4)
Requirement already satisfied: aiosignal>=1.1.2 in /usr/local/lib/python3.12/site-packages (from aiohttp>=3.8.1->toyota-na) (1.3.1)
Requirement already satisfied: attrs>=17.3.0 in /usr/local/lib/python3.12/site-packages (from aiohttp>=3.8.1->toyota-na) (24.2.0)
Requirement already satisfied: frozenlist>=1.1.1 in /usr/local/lib/python3.12/site-packages (from aiohttp>=3.8.1->toyota-na) (1.5.0)
Requirement already satisfied: multidict<7.0,>=4.5 in /usr/local/lib/python3.12/site-packages (from aiohttp>=3.8.1->toyota-na) (6.1.0)
Requirement already satisfied: yarl<2.0,>=1.12.0 in /usr/local/lib/python3.12/site-packages (from aiohttp>=3.8.1->toyota-na) (1.17.1)
Requirement already satisfied: cffi>=1.12 in /usr/local/lib/python3.12/site-packages (from cryptography>=35.0.0->toyota-na) (1.17.1)
Requirement already satisfied: pycparser in /usr/local/lib/python3.12/site-packages (from cffi>=1.12->cryptography>=35.0.0->toyota-na) (2.22)
Requirement already satisfied: idna>=2.0 in /usr/local/lib/python3.12/site-packages (from yarl<2.0,>=1.12.0->aiohttp>=3.8.1->toyota-na) (3.10)
Requirement already satisfied: propcache>=0.2.0 in /usr/local/lib/python3.12/site-packages (from yarl<2.0,>=1.12.0->aiohttp>=3.8.1->toyota-na) (0.2.0)
```

I checked my deps folder in my config and it's empty.

What do I need to do to get the required file in there?

**Comment by finity69x2** [comment]
Or maybe I'm battling a different issue?

here is the info from the HA integration page

![image](https://github.com/user-attachments/assets/b5e049d8-6b8c-4a97-a502-ce1578713b93)

**Comment by seang96** [comment]
@finity69x2 

You are experiecing issue #99 which was fixed in the latest update. 

@klein1jo 

That command if run as start command for container that sets the python path should fix it. It only sets the env for the command being ran. I suggest this way since it also includes the original Python path. You could try it in env variable of the container too and include the default Python path. I use Kubernetes so the configuration on how we run these are slightly different unfortunately.

**Comment by finity69x2** [comment]
You're right. Followed the instructions there ands all good again now. Thanks.

**Comment by seang96** [comment]
This probably should be brought up for an official fix in HASS image, I never went that route though since this is the only integration I experienced this with. It may be because HACs handles it differently vs a custom install?

**Comment by klein1jo** [comment]
It's 100% because HACS handles it differently.  All my HACS installed integrations broke with the 2024.12 container image and all for the same reason, no module found.  Maybe in the switch to Python 3.13 they accidentally removed an include for /config/deps?  I'll open an issue with the core project.

I'm using `podman create` statements for my containers.  In the meantime I took your lead and just added an environment variable to my hass create statement.

`-e PYTHONPATH="$PYTHONPATH:/config/deps"`

All my HACS installed integrations are working again.  Thank you @seang96.  Closing this issue this off as not an issue with this project.
